### DISCUSSION TITLE
Passing complex types such as a string from WasmEdge C SDK to Rust based Wasm Module

### DISCUSSION THREAD

**Discussion by GarethEgerton** [WasmEdge/WasmEdge#3342] [General]
Hi, I am trying to find an example of passing complex types such as string or byte array using the WasmEdge C Sdk through WasmEdge to my wasm module.

Currently I can see an example for Go and Rust here: [Pass complex parameters to WASM functions | WasmEdge Developer Guides](https://wasmedge.org/docs/embed/go/passing_data)

However there don't appear to be any examples of passing a string or byte array to a wasm module via the C Wasmedge API, just the fibonacci example which only passes integers.

There is an old example here: [wasm-cpp/TestWasmEdgeMain.cpp at 92f6538c5995d2dd6771592b91de622374ab0928 · nebula-hackathon-2021/wasm-cpp (github.com)](https://github.com/nebula-hackathon-2021/wasm-cpp/blob/92f6538c5995d2dd6771592b91de622374ab0928/TestWasmEdgeMain.cpp#L28) from another repo however it appears that the sdk has since been upgraded and this example no longer runs correctly.

Please, it would be much appreciated if anyone could advise me, or provide a working example of passing a string or byte array via C SDK to a Rust based wasm module?

Cheers,
Gareth

**Comment by q82419** [comment]
Hi,

Due to the limitation of WASM, the input and output can only be non-array types.
For passing the string/bytes, the examples above did the trick about implementing the function of allocation and deallocation of byte array.
Therefore, after compiling into WASM, users can invoke the function to allocate the array and get the offset.
You can try to use the [WasmEdge C SDK](https://wasmedge.org/docs/category/c-sdk-for-embedding-wasmedge) to invoke the exported that such of WASM functions from the self-made rust based WASM module.

For the other workaround, you can also try to use the `argc` and `argv` in WASI mode.

Thanks!

**Reply by GarethEgerton** [reply]
Thanks for this info :) it was very useful and I was able to create a basic example using the C SDK.
